Add 5.0 Actions Middleware change (#9943)
withastro/docs: src/content/docs/en/reference/modules/astro-actions.mdx

<Since v="4.15.0" />
</p>

The `message` property accepts a string. (e.g. "User must be logged in.")

### `getActionContext()`

<p>

> <Since v="5.0.0" />

Reviewer: nit, I think we can add the type here too:
```suggestion

**Type:** `(context: APIContext) => ActionMiddlewareContext`
<Since v="5.0.0" />
```

<Since v="4.15.0" />
</p>

The `message` property accepts a string. (e.g. "User must be logged in.")

### `getActionContext()`

<p>

**Type:** `(context: APIContext) => ActionMiddlewareContext`
<Since v="5.0.0" />
</p>

You may choose to customize how action results are handled. `getActionContext()` lets you programmatically get and set action results using middleware, allowing you to persist action results from HTML forms, gate action requests with added security checks, and more.

> Call `getActionContext()` from your middleware handler to retrieve information about inbound action requests. This function returns an `action` object with information about the request, and the `setActionResult()` and `serializeActionResult()` functions to programmically set the value returned by `Astro.getActionResult()`:

Reviewer: Sorry @bholmesdev, it seems I can't push directly to the branch: permissions issue... or I don't know how to do it. So I'm not sure making a PR targeting this branch is easier just for a small change. 😅 After that, I think it will be good for me unless there are other changes and it needs another proofread! 😉 

typo:
```suggestion
Call `getActionContext()` from your middleware handler to retrieve information about inbound action requests. This function returns an `action` object with information about the request, and the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`:
```

Reviewer 2: I got you!

> You may choose to customize how action results are handled. `getActionContext()` lets you programmatically get and set action results using middleware, allowing you to persist action results from HTML forms, gate action requests with added security checks, and more.
> 
> Call `getActionContext()` from your middleware handler to retrieve information about inbound action requests. This function returns an `action` object with information about the request, and the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`:

Reviewer 2: ```suggestion
`getActionContext()` is a function called from your middleware handler to retrieve information about inbound action requests. 

This function returns an `action` object with information about the request, and the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`.

`getActionContext()` lets you programmatically get and set action results using middleware, allowing you to persist action results from HTML forms, gate action requests with added security checks, and more.
```
For the reference page, we'd tend to go boring definition stuff first.

```ts title="src/middleware.ts" {5}
import { defineMiddleware } from 'astro:middleware';
import { getActionContext } from 'astro:actions';

export const onRequest = defineMiddleware(async (context, next) => {
  const { action, setActionResult, serializeActionResult } = getActionContext(context);
  if (action?.calledFrom === 'form') {
    const result = await action.handler();
    setActionResult(action.name, serializeActionResult(result));
  }
  return next();
});
```

#### `action`

<p>

**Type:** `{ calledFrom: 'rpc' | 'form', name: string, handler: () => Promise<SafeResult<any, any>> } | undefined`
</p>

> `action` exposes information about an inbound action request. When an action is called, `action` will contain the action name, handler, and whether the action was called from an client-side RPC function (example: `actions.newsletter()`) or an HTML form action.

Reviewer 2: ```suggestion
`action` is an object containing information about an inbound action request.

 It is available from `getActionContext()`, and provides the action name, handler, and whether the action was called from an client-side RPC function (e.g. `actions.newsletter()`) or an HTML form action.
```
Again, just trying to make this a bit more boring and referencey!

```ts title="src/middleware.ts" {6}
import { defineMiddleware } from 'astro:middleware';
import { getActionContext } from 'astro:actions';

export const onRequest = defineMiddleware(async (context, next) => {
  const { action, setActionResult, serializeActionResult } = getActionContext(context);
  if (action?.calledFrom === 'rpc' && action.name.startsWith('private')) {
    // Check for a valid session token
  }
  // ...
});
```

#### `setActionResult()`

<p>

**Type:** `(actionName: string, actionResult: SerializedActionResult) => void`
</p>

> `setActionResult()` will programmatically set the value returned by `Astro.getActionResult()`. This is useful when calling actions from an HTML form to persist and load results from a session.

Author: Agreed. Also don't think this is correct:

> passed the action name and the function to serialize the action result to JSON.

You don't pass a _function,_ you pass a serialized action result!

<ReadMore>See the [advanced sessions guide](/en/guides/actions/#advanced-persist-action-results-with-a-session) for a sample implementation using Netlify Blob.</ReadMore>

#### `serializeActionResult()`

<p>

**Type:** `(result: SafeResult<any, any>) => SerializedActionResult`
</p>

`serializeActionResult()` will serialize an action result to JSON for persistence. This is required to properly handle non-JSON return values like `Map` or `Date` as well as the `ActionError` object. Be sure to call this function when serializing an action result to be passed to `setActionResult()`:

```ts title="src/middleware.ts" {8}
import { defineMiddleware } from 'astro:middleware';
import { getActionContext } from 'astro:actions';

export const onRequest = defineMiddleware(async (context, next) => {
  const { action, setActionResult, serializeActionResult } = getActionContext(context);
  if (action) {
    const result = await action.handler();
    setActionResult(action.name, serializeActionResult(result));
  }
  // ...
});
```

#### `deserializeActionResult()`

<p>

**Type:** `(result: SerializedActionResult) => SafeResult<any, any>`
</p>

> `deserializeActionResult()` will deserialize an action result serialized using `serializeActionResult()`.

Reviewer: Ah, good point! Perhaps we can use the JSDoc description to be more precise about the usefulness of this method: `to access data and error objects.`?

> `deserializeActionResult()` will deserialize an action result serialized using `serializeActionResult()`.

Author: Agreed with you Armand! I prefer describing the usefulness over simply calling it a reversal function. I'll commit an updated draft here. By JSDoc, do you mean a change in the core code?

> `deserializeActionResult()` will deserialize an action result serialized using `serializeActionResult()`.

Reviewer: @bholmesdev  Oh no, I was just saying that I took this excerpt from the existing JSDoc. I wasn't suggesting modifying the JSDoc! 😉
